usr/src/uts/common/fs/zfs/sys/spa.h

```diff
@@ -326,7 +327,8 @@ typedef enum bp_embedded_type {
 typedef struct blkptr {
 	dva_t		blk_dva[SPA_DVAS_PER_BP]; /* Data Virtual Addresses */
 	uint64_t	blk_prop;	/* size, compression, type, etc	    */
-	uint64_t	blk_pad[2];	/* Extra space for the future	    */
+	uint32_t	blk_iv[3];	/* IV for block level encryption    */
+	uint32_t	blk_pad;	/* Extra space for the future	    */
```

Comment:
> My point was, as long as ZFS encryption uses a cipher that provides message integrity, then there should be no need to also have a data "checksum" for data that's encrypted.

There is a need, actually. The message integrity given by the encryption algorithms we use is only useful if we have the key loaded. Without the regular checksum, ZFS can't detect any faults in a dataset without the keys loaded. This means that admin scrubbing, resilvering, etc all can't happen anymore.